### PR TITLE
Add column properties to SQL parser

### PR DESCRIPTION
**Pull request by MartinWeindel**
The Kudu connector needs to specify additional properties for columns on creating tables (or altering a table by adding a new column).
Following the existing syntax for table properties, this PR adds column properties to the parser and all related infrastructure to manage them similar to table properties.
Formatting SQL code for `SHOW CREATE TABLE` has also been updated accordingly.

Example for new column properties syntax from the Kudu connector:
```sql
CREATE TABLE kudu.default.sample (
  serialno varchar WITH ( primary_key = true )
  col1 varchar WITH ( encoding = 'dictionary', compression = 'LZ4' ),  
  col2 varchar WITH ( nullable = true )
);
```
The concrete column properties are defined in the connectors.

### CODE REVIEW

**Comment by dain** [comment]
Isn't there standard syntax for primary key and nullable?  The others make sense to me.

**Comment by MartinWeindel** [comment]
It is not my focus to make a large rework of the Presto SQL parser.

Adding standard ANSI column flags like `NULLABLE` and `PRIMARY KEY` for `CREATE TABLE` has much larger impact as then all connector would have to check if these flags are set and reject them if they are not supported.
And there is no reason to support it later if somebody wants to implement such ANSI SQL standard features.

**Comment by kokosing** [comment]
> all connector would have to check if these flags are set and reject them if they are not supported.

I think it would be nice to have something like this implemented here as well. In case when someone uses column property that given connector does not support I would expect exception to be thrown. Presto behaves like that today, for example if you use wrong property in `config.properties` Presto won't start. The same for connector properties. However I do not know how it looks like with table properties.

**Comment by electrum** [comment]
@kokosing this already works for schema and table properties (and this as well since it uses the same code)

Standard proprties like “not null” are unrelated to this PR.

File: presto-main/src/main/java/com/facebook/presto/execution/CreateTableTask.java [hunk 1]
```diff
@@ -115,7 +119,23 @@ public String explain(CreateTable statement, List<Expression> parameters)
                 if (columns.containsKey(name)) {
                     throw new SemanticException(DUPLICATE_COLUMN_NAME, column, "Column name '%s' specified more than once", column.getName());
                 }
-                columns.put(name, new ColumnMetadata(name, type, column.getComment().orElse(null), false));
+
+                Map<String, Object> columnProperties;
+                if (!column.getProperties().isEmpty()) {
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/execution/CreateTableTask.java]: What `column.getProperties().isEmpty()` but connector requires a property to be set? Now `ColumnPropertyManager` does not have a chance to throw an exception.

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java [hunk 2]
```diff
@@ -453,6 +437,41 @@ protected Node visitShowCreate(ShowCreate node, Void context)
             throw new UnsupportedOperationException("SHOW CREATE only supported for tables and views");
         }
 
+        private List<Property> buildProperties(Object objectName, StandardErrorCode errorCode,
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java]: put eveyry parameter into separate line like:
```
ReturnType methodName(
       ParameterType1 pameter1,
       ParameterType2 pameter2,
       ParameterType3 pameter3,
       ParameterType4 pameter4)

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java [hunk 3]
```diff
@@ -453,6 +437,41 @@ protected Node visitShowCreate(ShowCreate node, Void context)
...
+                if (value == null) {
+                    throw new PrestoException(errorCode, format("Property %s for table %s cannot have a null value", propertyName, objectName));
+                }
+
+                PropertyMetadata<?> property = allProperties.get(propertyName);
+                if (!Primitives.wrap(property.getJavaType()).isInstance(value)) {
+                    throw new PrestoException(errorCode, format(
+                            "Property %s for table %s should have value of type %s, not %s",
+                            propertyName,
+                            objectName,
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java]: Using this `objectName` here is a bit hacky. It is not obvious (from reading the method body) that this method is called for column properties as well. Maybe instead of passing just `Object objectName` (`QualifiedColumnName `) you could pass an `Optional<String> columnName` together with `QualifiedObjectName objectName`.

File: presto-parser/src/main/java/com/facebook/presto/sql/SqlFormatter.java [hunk 4]
```diff
@@ -924,9 +939,7 @@ protected Void visitAddColumn(AddColumn node, Integer indent)
             builder.append("ALTER TABLE ")
                     .append(node.getName())
                     .append(" ADD COLUMN ")
-                    .append(node.getColumn().getName())
-                    .append(" ")
```
Comment (kokosing) [inline presto-parser/src/main/java/com/facebook/presto/sql/SqlFormatter.java]: It looks like you fixed a bug here, can you add a test for that (possibly extract as separate commit). 

Bug is about adding a column with `COMMENT`, old code was losing `COMMENT` here.

File: presto-parser/src/main/java/com/facebook/presto/sql/SqlFormatter.java [hunk 5]
```diff
@@ -837,23 +834,32 @@ protected Void visitCreateTable(CreateTable node, Integer indent)
...
                             formatExpression(element.getName(), parameters) + " = " +
                             formatExpression(element.getValue(), parameters))
-                    .collect(joining(",\n"));
+                    .collect(joining(delimiter));
 
-            return "\nWITH (\n" + propertyList + "\n)";
+            if (multiLine) {
+                return "\nWITH (\n" + propertyList + "\n)";
+            }
+            else {
```
Comment (electrum) [inline presto-parser/src/main/java/com/facebook/presto/sql/SqlFormatter.java]: Nit: the "else" is redundant since the "if' branch returns unconditionally

File: presto-parser/src/main/java/com/facebook/presto/sql/SqlFormatter.java [hunk 6]
```diff
@@ -837,23 +834,32 @@ protected Void visitCreateTable(CreateTable node, Integer indent)
...
             }
 
-            builder.append(formatProperties(node.getProperties()));
+            builder.append(formatProperties(node.getProperties(), true));
 
             return null;
         }
 
-        private String formatProperties(List<Property> properties)
+        private String formatProperties(List<Property> properties, boolean multiLine)
```
Comment (electrum) [inline presto-parser/src/main/java/com/facebook/presto/sql/SqlFormatter.java]: I'm thinking this would be easier to read if we have two methods `formatPropertiesOneLine` and `formatPropertiesMultiLine`. There will be a bit of code duplication, but each variant will be simpler.

File: presto-spi/src/main/java/com/facebook/presto/spi/ColumnMetadata.java [hunk 7]
```diff
@@ -51,6 +60,7 @@ public ColumnMetadata(String name, Type type, String comment, String extraInfo,
         this.comment = comment;
         this.extraInfo = extraInfo;
         this.hidden = hidden;
+        this.properties = (properties != null && !properties.isEmpty()) ? Collections.unmodifiableMap(new LinkedHashMap<>(properties)) : Collections.emptyMap();
```
Comment (electrum) [inline presto-spi/src/main/java/com/facebook/presto/spi/ColumnMetadata.java]: The Presto code base generally rejects nulls rather than being null-friendly, so we should not allow null here. Users should instead pass an empty collection. For example, `ConnectorTableMetadata` does not accept null properties.

Nit: static import `unmodifiableMap` and `emptyMap` for readability, as the method names are self-descriptive without the class prefix

File: presto-main/src/main/java/com/facebook/presto/execution/AddColumnTask.java [hunk 8]
```diff
@@ -79,7 +87,22 @@ public String getName()
             throw new SemanticException(COLUMN_ALREADY_EXISTS, statement, "Column '%s' already exists", element.getName());
         }
 
-        ColumnMetadata column = new ColumnMetadata(element.getName().getValue(), type, element.getComment().orElse(null), false);
+        Map<String, Object> columnProperties;
+        if (!element.getProperties().isEmpty()) {
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/execution/AddColumnTask.java]: We shouldn't need to special case empty here

**Comment by electrum** [comment]
Overall this looks good. I'll do a final review and merge after the release goes out.

**Comment by MartinWeindel** [comment]
@electrum Are there any chances to finalize this pull request in the near future? I'm depending on this PR for further work on the Kudu connector.

**Comment by kokosing** [comment]
Thanks!
